File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
from collections import deque
from sys import stdin

n, m = map(int, stdin.readline().split())
ladders = {}
snakes = {}
for _ in range(n):
    s, d = map(int, stdin.readline().split())
    ladders[s] = d

max_v = 0
for _ in range(m):
    s, d = map(int, stdin.readline().split())
    snakes[s] = d


def bfs():
    global ladders, snakes
    dx = [1, 2, 3, 4, 5, 6]
    que = deque([[1, 0]])
    visit = {}
    while que:
        x, cnt = que.popleft()
        for i in range(6):
            nx = x + dx[i]
            if nx == 100:
                return cnt + 1
            if ladders.get(nx) is not None:
                visit[nx] = True
                nx = ladders[nx]
            if snakes.get(nx) is not None:
                visit[nx] = True
                nx = snakes[nx]
            if nx == 100:
                return cnt + 1
            if visit.get(nx) is None:
                visit[nx] = True
                que.append([nx, cnt + 1])


print(bfs())

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());

        int[] ladders = new int[101];
        int[] snakes = new int[101];

        for(int i=0 ; i<101; i++){
            ladders[i] = i;
            snakes[i] = i;
        }

        for(int i=0 ; i<n ; i++){
            tokenizer = new StringTokenizer(br.readLine());
            int sour = Integer.parseInt(tokenizer.nextToken());
            int des = Integer.parseInt(tokenizer.nextToken());

            ladders[sour] = des;
        }

        for(int i=0 ; i<m ; i++){
            tokenizer = new StringTokenizer(br.readLine());
            int sour = Integer.parseInt(tokenizer.nextToken());
            int des = Integer.parseInt(tokenizer.nextToken());

            snakes[sour] = des;
        }

        bfs(ladders, snakes);
    }

    private static void bfs(final int[] ladders, final int[] snakes){
        Queue<int[]> que = new LinkedList<>();
        que.add(new int[]{1, 0});

        boolean[] visit = new boolean[101];
        visit[1] = true;

        while(!que.isEmpty()){
            int[] cur = que.poll();
            int point = cur[0], cnt = cur[1];

            for(int i=1 ; i<=6 ; i++){
                int nextPoint = point + i;

                if(nextPoint == 100){
                    System.out.println(cnt + 1);
                    return;
                }

                if(visit[nextPoint])continue;
                visit[nextPoint] = true;

                if(snakes[nextPoint] != nextPoint){
                    nextPoint = snakes[nextPoint];
                }else{
                    nextPoint = ladders[nextPoint];
                }

                visit[nextPoint] = true;
                que.add(new int[]{nextPoint, cnt + 1});
            }
        }
    }
}